src/node/config.ts

```diff
@@ -627,106 +624,75 @@ async function loadConfigFromBundledFile(
-      plugin.configureBuild || []
-    ),
-    vueCompilerOptions: {
-      ...config.vueCompilerOptions,
-      ...plugin.vueCompilerOptions
-    },
-    vueTransformAssetUrls: mergeAssetUrlOptions(
-      config.vueTransformAssetUrls,
-      plugin.vueTransformAssetUrls
-    ),
-    vueTemplatePreprocessOptions: {
-      ...config.vueTemplatePreprocessOptions,
-      ...plugin.vueTemplatePreprocessOptions
-    },
-    vueCustomBlockTransforms: {
-      ...config.vueCustomBlockTransforms,
-      ...plugin.vueCustomBlockTransforms
-    },
-    rollupInputOptions: mergeObjectOptions(
-      config.rollupInputOptions,
-      plugin.rollupInputOptions
-    ),
-    rollupOutputOptions: mergeObjectOptions(
-      config.rollupOutputOptions,
-      plugin.rollupOutputOptions
-    ),
-    enableRollupPluginVue:
-      config.enableRollupPluginVue || plugin.enableRollupPluginVue
+function mergePlugin(config: ResolvedConfig, plugin: Plugin) {
+  if (typeof plugin === 'function') {
+    plugin = plugin(config)
+  }
+  for (const key in plugin) {
+    let value = (plugin as any)[key]
+    if (value == null) {
+      continue
+    }
+    const oldValue = (config as any)[key]
+    // normalize the asset url options before merging
+    if (key === 'vueTransformAssetUrls') {
+      if (isPlainObject(oldValue)) {
+        if (isPlainObject(value))
+          Object.assign(oldValue, normalizeAssetUrlOptions(value))
+        continue // ignore boolean when object exists
+      }
+    }
+    // let `true` take precedence
+    else if (key === 'enableRollupPluginVue' && !value) {
+      continue
+    }
+    // prefer merging into an existing array
+    else if (Array.isArray(value)) {
+      if (Array.isArray(oldValue)) {
+        value.forEach((value) => oldValue.push(value))
+        continue
+      }
+      // shallow clone to assume ownership
+      value = [...value]
+    }
+    // prefer merging into an existing object (shallowly)
+    else if (isPlainObject(value)) {
+      if (isPlainObject(oldValue)) {
+        const overrides: any = {}
+        if (/^rollup(In|Out)putOptions$/.test(key)) {
+          overrides.plugins = mergeArrays(
+            (oldValue as any).plugins,
+            (value as any).plugins
+          )
+          if (key === 'rollupInputOptions')
+            overrides.acornInjectPlugins = mergeArrays(
+              (oldValue as any).acornInjectPlugins,
+              (value as any).acornInjectPlugins
+            )
+        }
+        Object.assign(oldValue, value, overrides)
+        continue
+      }
+      // shallow clone to assume ownership
+      value = { ...value }
+    }
+    // overwrite when merging is unnecessary
+    ;(config as any)[key] = value
   }
 }
 
-function mergeAssetUrlOptions(
-  to: SFCTemplateCompileOptions['transformAssetUrls'],
-  from: SFCTemplateCompileOptions['transformAssetUrls']
-): SFCTemplateCompileOptions['transformAssetUrls'] {
-  if (from === true) {
-    return to
```

Comment:
@yyx990803 This line implies that an existing value of `vueTransformAssetUrls: false` should take precedence over a new value of `vueTransformAssetUrls: true`. Was this intended?